services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/HankealueService.kt

```diff
@@ -0,0 +1,157 @@
+import fi.hel.haitaton.hanke.domain.NewGeometriat
+import fi.hel.haitaton.hanke.domain.NewHankealue
+import fi.hel.haitaton.hanke.domain.SavedHankealue
+import fi.hel.haitaton.hanke.geometria.Geometriat
+import fi.hel.haitaton.hanke.geometria.GeometriatService
+import fi.hel.haitaton.hanke.tormaystarkastelu.TormaystarkasteluLaskentaService
+import fi.hel.haitaton.hanke.tormaystarkastelu.TormaystarkasteluTulosEntity
+import org.geojson.Feature
+import org.geojson.FeatureCollection
+import org.springframework.stereotype.Service
+
+/*
+ * Helper for mapping and sorting data to existing collections.
+ *
+ * Example usage:
+ * mergeDataInto(hanke.alueet, entity.listOfHankeAlueet) { source, target ->
+ *    copyNonNullHankealueFieldsToEntity(hanke, source, target)
+ * }
+ * - Transforms data from hanke.alueet into entity.listOfHankeAlueet
+ * - Does the transformation with the last lambda
+ *
+ * Source list is not modified. Target container is sorted by order of source container.
+ *
+ * converter takes additional parameter for existing Target object in case it exists in target collection.
+ *
+ */
+private fun <Source : HasId<Int>, Target : HasId<Int>> mergeDataInto(
+    source: List<Source>,
+    target: MutableList<Target>,
+    converterFn: (Source, Target?) -> Target
+) {
+    // Existing data is collected for mapping
+    val (targetIdentified, targetRest) = target.partition { it.id != null }
+    val targetMap = targetIdentified.associateBy { it.id!! }
+
+    // Target is overwritten with merged and new data from source
+    target.clear()
+    target.addAll(source.map { converterFn(it, targetMap[it.id]) })
+    target.addAll(targetRest)
+}
+
+@Service
+class HankealueService(
+    private val geometriatService: GeometriatService,
+    private val tormaystarkasteluService: TormaystarkasteluLaskentaService
+) {
+
+    fun mergeAlueetToHanke(incoming: List<SavedHankealue>, existingHanke: HankeEntity) {
+        mergeDataInto(incoming, existingHanke.alueet) { source, target ->
+            copyNonNullHankealueFieldsToEntity(existingHanke.hankeTunnus, source, target)
+        }
+        existingHanke.alueet.forEach { it.hanke = existingHanke }
+    }
+
+    /** Map by area geometry id to area geometry data. */
+    fun geometryMapFrom(alueet: List<HankealueEntity>): Map<Int, Geometriat?> =
+        alueet
+            .mapNotNull { it.geometriat }
+            .associateBy({ it }, { geometriatService.getGeometriat(it) })
+
+    fun copyNonNullHankealueFieldsToEntity(
+        hankeTunnus: String,
+        source: Hankealue,
+        target: HankealueEntity?
+    ): HankealueEntity {
+        val result = target ?: HankealueEntity()
+
+        // Assuming the incoming date, while being zoned date and time, is in UTC and time value can
+        // be simply dropped here.
+        // Note, .toLocalDate() does not do any time zone conversion.
+        source.haittaLoppuPvm?.let { result.haittaLoppuPvm = source.haittaLoppuPvm?.toLocalDate() }
```

Comment:
Would it make sense to use it for all of these? 

`source.haittaLoppuPvm?.let { result.haittaLoppuPvm = it.toLocalDate() }`

Reply:
I changed them.